File: 卡码/代码随想录/Java/src/com/kurumi/codeThoughts/回溯/_03_子集/LeetCode_0078.java
package com.kurumi.codeThoughts.回溯._03_子集;

import java.util.ArrayList;
import java.util.Deque;
import java.util.LinkedList;
import java.util.List;

/**
 * @Author kurumi
 * @Date 2026-07-03 上午11:52:20
 * @Description 78. 子集
 */
public class LeetCode_0078 {
    static class Solution {
        Deque<Integer> path = new LinkedList<>();
        List<List<Integer>> res = new ArrayList<>();

        public void backtracking(int[] nums, int index) {
            res.add(new ArrayList<>(path));
            if (index >= nums.length)
                return;
            for (int i = index; i < nums.length; i++) {
                path.add(nums[i]);
                backtracking(nums, i + 1);
                path.removeLast();
            }
        }

        public List<List<Integer>> subsets(int[] nums) {
            backtracking(nums, 0);
            return res;
        }
    }
}

File: 卡码/代码随想录/C++/回溯/03-子集/LeetCode_0078.cpp
//78. 子集
#include <iostream>
#include <vector>
using namespace std;

class Solution {
	vector<int> path;
	vector<vector<int>> res;
public:
	void backtracking(vector<int>& nums,int index){
		res.push_back(path);
		if(index>=nums.size()) return;
		for(int i=index;i<nums.size();i++){
			path.push_back(nums[i]);
			backtracking(nums,i+1);
			path.pop_back();
		}
	}
	vector<vector<int>> subsets(vector<int>& nums) {
		backtracking(nums,0);
		return res;
	}
};

int main(){
  system("pause");
  return 0;
}
